port/unix/omrsysinfo.c

```diff
@@ -2004,6 +2013,75 @@ omrsysinfo_get_OS_version(struct OMRPortLibrary *portLibrary)
 					rc = sysctlbyname(KERN_OSPRODUCTVERSION, buffer, &resultSize, NULL, 0);
 				}
 				buffer[resultSize] = '\0';
+
+				/* 10.16 is the version returned when in a backwards compatibility context in OSX versions 11+
+				* so we check the plink symlink as that bypasses the compatibility context and returns the
+				* real version info.
+				*/
+				if (0 == strncmp(buffer, "10.16", 5)) {
+					portLibrary->mem_free_memory(portLibrary, buffer);
+					buffer = NULL;
+					len = 0;
+					portLibrary->mem_free_memory(portLibrary, allocatedFileBuffer);
+					allocatedFileBuffer = NULL;
+
+					intptr_t plistFile = portLibrary->file_open(portLibrary, COMPAT_PLIST_LINK, EsOpenRead, 0444);
+					if (plistFile >= 0) {
+						char myBuffer[1024];  /* the file size is ~500 bytes, 525 on a 11.3 */
+						char *fileBuffer = myBuffer;
+						size_t fileLen = 0;
+						rc = portLibrary->file_flength(portLibrary, plistFile);
+						if (rc > 0) {
+							fileLen = (size_t) rc;
+							if (fileLen >= sizeof(myBuffer)) { /* leave space for a terminating null */
+								allocatedFileBuffer = portLibrary->mem_allocate_memory(portLibrary, fileLen + 1,
+									OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
+								fileBuffer = allocatedFileBuffer;
+							} else {
+								fileBuffer = myBuffer;
+							}
+						}
+
+						rc = 0;
+						if (NULL != fileBuffer) {
+							char *readPtr = fileBuffer;
+							size_t bytesRemaining = fileLen;
+							while (bytesRemaining > 0) {
+								intptr_t result = portLibrary->file_read(portLibrary, plistFile, readPtr, bytesRemaining);
+								if (result > 0) {
+									bytesRemaining -= result;
+									readPtr += result;
+								} else {
+									rc = -1; /* error or unexpected EOF */
+									break;
+								}
+							}
+							if (0 == rc) {
+								fileBuffer[fileLen] = '\0';
+								readPtr = strstr(fileBuffer, PRODUCT_VERSION_ELEMENT);
+								if (NULL != readPtr) {
+									readPtr = strstr(readPtr + sizeof(PRODUCT_VERSION_ELEMENT) - 1, STRING_OPEN_TAG);
+								}
+								if (NULL != readPtr) {
+									versionString = readPtr + sizeof(STRING_OPEN_TAG) - 1;
+									readPtr = strstr(versionString, STRING_CLOSE_TAG);
+								}
+								if (NULL != readPtr) {
+									*readPtr = '\0';
+									resultSize = readPtr - versionString;
+
+									len = resultSize + 1;
+									buffer = portLibrary->mem_allocate_memory(portLibrary, len,
+										OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
+									if (NULL != buffer) {
+										strncpy(buffer, versionString, len);
```

Comment:
Rather than write the NUL to `readPtr` above, I suggest this should use `memcpy()`:
```
memcpy(buffer, versionString, resultSize);
buffer[resultSize] = '\0';
```